mesonbuild/interpreter/interpreter.py

```diff
@@ -687,10 +687,19 @@ def func_declare_dependency(self, node, args, kwargs):
         for d in deps:
             if not isinstance(d, dependencies.Dependency):
                 raise InterpreterException('Invalid dependency')
+
+        dep_err = 'declare_dependency: Entries in "{}" may only be self-built targets, external dependencies (including libraries) must go to "dependencies".'
+
         for l in libs:
             if isinstance(l, dependencies.Dependency):
-                raise InterpreterException('''Entries in "link_with" may only be self-built targets,
-external dependencies (including libraries) must go to "dependencies".''')
+                raise InvalidArguments(dep_err.format('link_with'))
+        for l in libs_whole:
+            if isinstance(l, dependencies.Dependency):
+                raise InvalidArguments(dep_err.format('link_whole'))
+            if isinstance(l, build.SharedLibrary):
+                raise InvalidArguments('declare_dependency: SharedLibrary objects are not allowed in "link_whole"')
+            if isinstance(l, (build.CustomTarget, build.CustomTargetIndex)) and l.links_dynamically():
+                raise InvalidArguments('declare_dependency: CustomTarget and CustomTargetIndex returning shared libaries are not allowed in "link_whole"')
```

Comment:
The previous commit treats both of these the same and just calls them "shared libraries" without respect to what represented type it is. We should probably either change this to not distinguish them, or change the previous commit to distinguish them.

Reply:
I think being more specific is better than being less, so I'll convert the previous case